### ISSUE TITLE
Exported onnx graph is invalid when using a traced module with Conv and BatchNorm inside a scripted module with If

### ISSUE THREAD

**Issue by waihokwok** [pytorch/pytorch#68724]
## 🐛 Bug

When using a traced module with Conv and BatchNorm inside a scripted module with If, the output onnx graph is invalid.

### Code
```python
import numpy as np
import torch
from torch import nn
import onnx
import onnxruntime


dummy_in = torch.ones(1, 3, 100, 100)


class ConvBlock(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(3, 16, 3, padding=1, bias=False)
        self.norm = nn.BatchNorm2d(16)

    def forward(self, x: torch.Tensor):
        return self.norm(self.conv(x))


class TraceInScriptWithIf(nn.Module):
    def __init__(self):
        super().__init__()
        self.traced_conv_one_block = torch.jit.trace(  # type:ignore
            ConvBlock().eval(), (dummy_in))
        print(self.traced_conv_one_block.graph)
        self.traced_conv_two = torch.jit.trace(  # type:ignore
            ConvBlock().eval(), (dummy_in))
        print(self.traced_conv_two.graph)

    def forward(self, x):
        if x.sum() > 0:
            return self.traced_conv_one_block(x)
        return self.traced_conv_two(x)


scripted_control_flow = torch.jit.script(TraceInScriptWithIf())  # type:ignore
print(scripted_control_flow.graph)
dummy_out = scripted_control_flow(dummy_in)
# Convert onnx
onnx_model_name = 'script_in_trace_with_if.onnx'
torch.onnx.export(
    scripted_control_flow,
    (dummy_in),
    onnx_model_name,
    do_constant_folding=True,
    opset_version=11,
    input_names=["input"],
    output_names=["output"],
    dynamic_axes={
        "input": {0: "n", 2: "h", 3: "w"},
        "output": {0: "n", 2: "h", 3: "w"},
    },
)
# Print onnx graph
onnx_model = onnx.load(onnx_model_name)
print(onnx.helper.printable_graph(onnx_model.graph))  # type: ignore

# Test Onnx Graph


def to_numpy(tensor: torch.Tensor) -> np.ndarray:
    if tensor.requires_grad:
        return tensor.detach().cpu().numpy()
    else:
        return tensor.cpu().numpy()


def all_close(tensor: torch.Tensor, ndarray: np.ndarray) -> str:
    return 'OK' if np.allclose(to_numpy(tensor), ndarray, rtol=1e-03, atol=1e-05) else 'not OK'


ort_session = onnxruntime.InferenceSession(
    onnx_model_name, providers=["CUDAExecutionProvider"],
)

ort_dummy_in_dict = {"input": to_numpy(dummy_in)}
ort_dummy_out_list = ort_session.run(None, ort_dummy_in_dict)
ort_dummy_out = ort_dummy_out_list[0]
print('Check ort inference')
print('ort_zeros_out', all_close(dummy_out, ort_dummy_out))
```

### Output
```
graph(%self.1 : __torch__.ConvBlock,
      %x : Float(1, 3, 100, 100, strides=[30000, 10000, 100, 1], requires_grad=0, device=cpu)):
  %norm : __torch__.torch.nn.modules.batchnorm.BatchNorm2d = prim::GetAttr[name="norm"](%self.1)
  %conv : __torch__.torch.nn.modules.conv.Conv2d = prim::GetAttr[name="conv"](%self.1)
  %52 : Tensor = prim::CallMethod[name="forward"](%conv, %x)
  %53 : Tensor = prim::CallMethod[name="forward"](%norm, %52)
  return (%53)

graph(%self.1 : __torch__.___torch_mangle_5.ConvBlock,
      %x : Float(1, 3, 100, 100, strides=[30000, 10000, 100, 1], requires_grad=0, device=cpu)):
  %norm : __torch__.torch.nn.modules.batchnorm.___torch_mangle_4.BatchNorm2d = prim::GetAttr[name="norm"](%self.1)
  %conv : __torch__.torch.nn.modules.conv.___torch_mangle_3.Conv2d = prim::GetAttr[name="conv"](%self.1)
  %52 : Tensor = prim::CallMethod[name="forward"](%conv, %x)
  %53 : Tensor = prim::CallMethod[name="forward"](%norm, %52)
  return (%53)

graph(%self : __torch__.TraceInScriptWithIf,
      %x.1 : Tensor):
  %3 : NoneType = prim::Constant()
  %5 : int = prim::Constant[value=0]() # convert_trace_in_script_with_if.py:32:21
  %4 : Tensor = aten::sum(%x.1, %3) # convert_trace_in_script_with_if.py:32:11
  %6 : Tensor = aten::gt(%4, %5) # convert_trace_in_script_with_if.py:32:11
  %8 : bool = aten::Bool(%6) # convert_trace_in_script_with_if.py:32:11
  %31 : Tensor = prim::If(%8) # convert_trace_in_script_with_if.py:32:8
    block0():
      %traced_conv_one_block : __torch__.ConvBlock = prim::GetAttr[name="traced_conv_one_block"](%self)
      %11 : Tensor = prim::CallMethod[name="forward"](%traced_conv_one_block, %x.1) # convert_trace_in_script_with_if.py:33:19
      -> (%11)
    block1():
      %traced_conv_two.1 : __torch__.___torch_mangle_5.ConvBlock = prim::GetAttr[name="traced_conv_two"](%self)
      %29 : Tensor = prim::CallMethod[name="forward"](%traced_conv_two.1, %x.1) # convert_trace_in_script_with_if.py:34:15
      -> (%29)
  return (%31)

graph torch-jit-export (
  %input[FLOAT, nx3xhxw]
) initializers (
  %21[FLOAT, 16x3x3x3]
  %24[FLOAT, 16x3x3x3]
) {
  %11 = ReduceSum[keepdims = 0](%input)
  %12 = Constant[value = <Scalar Tensor []>]()
  %13 = Greater(%11, %12)
  %14 = Cast[to = 9](%13)
  %output = If[else_branch = <graph torch-jit-export2>, then_branch = <graph torch-jit-export1>](%14)
  return %output
}

graph torch-jit-export1 (
  %22[FLOAT, 16]
) {
  %20 = Conv[dilations = [1, 1], group = 1, kernel_shape = [3, 3], pads = [1, 1, 1, 1], strides = [1, 1]](%input, %21, %22)
  return %20
}

graph torch-jit-export2 (
  %25[FLOAT, 16]
) {
  %23 = Conv[dilations = [1, 1], group = 1, kernel_shape = [3, 3], pads = [1, 1, 1, 1], strides = [1, 1]](%input, %24, %25)
  return %23
}
Traceback (most recent call last):
  File "convert_trace_in_script_with_if.py", line 73, in <module>
    ort_session = onnxruntime.InferenceSession(
  File "/home/whkwok/miniconda3/envs/pytorch/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 324, in __init__
    self._create_inference_session(providers, provider_options, disabled_optimizers)
  File "/home/whkwok/miniconda3/envs/pytorch/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 358, in _create_inference_session
    sess = C.InferenceSession(session_options, self._model_path, True, self._read_config_from_model)
onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Load model from script_in_trace_with_if.onnx failed:Node (If_4) Op (If) [TypeInferenceError] Graph attribute inferencing failed: Size mismatch validating subgraph inputs. Got 0 inputs but subgraph has 1 inputs and requires 1 inputs. Either provide all subgraph inputs, or just the required inputs.
```

I suspect that the issues comes from the jit optimization passes that try to fuse Conv and BatchNorm. The generated subgraphs torch-jit-export1 and torch-jit-export2 should not take any inputs according to onnx specification. %22 and %25, which is the biases of BatchNorms, should be treated as initializers instead.

## Workaround
Adding zero between Conv and BatchNorm can stop the fusing and the exported graph become valid (but unoptimized).
### Code
```python
# Add a zero between Conv and BatchNorm
class ConvBlock(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(3, 16, 3, padding=1, bias=False)
        self.norm = nn.BatchNorm2d(16)

    def forward(self, x: torch.Tensor):
        return self.norm(self.conv(x) + 0) # stop fusing Conv and BatchNorm
```
### Output
```
graph(%self.1 : __torch__.ConvBlock,
      %x : Float(1, 3, 100, 100, strides=[30000, 10000, 100, 1], requires_grad=0, device=cpu)):
  %norm : __torch__.torch.nn.modules.batchnorm.BatchNorm2d = prim::GetAttr[name="norm"](%self.1)
  %conv : __torch__.torch.nn.modules.conv.Conv2d = prim::GetAttr[name="conv"](%self.1)
  %55 : Tensor = prim::CallMethod[name="forward"](%conv, %x)
  %36 : Long(requires_grad=0, device=cpu) = prim::Constant[value={0}]() # convert_trace_in_script_with_if.py:18:0
  %37 : int = prim::Constant[value=1]() # convert_trace_in_script_with_if.py:18:0
  %input : Float(1, 16, 100, 100, strides=[160000, 10000, 100, 1], requires_grad=1, device=cpu) = aten::add(%55, %36, %37) # convert_trace_in_script_with_if.py:18:0
  %56 : Tensor = prim::CallMethod[name="forward"](%norm, %input)
  return (%56)

graph(%self.1 : __torch__.___torch_mangle_5.ConvBlock,
      %x : Float(1, 3, 100, 100, strides=[30000, 10000, 100, 1], requires_grad=0, device=cpu)):
  %norm : __torch__.torch.nn.modules.batchnorm.___torch_mangle_4.BatchNorm2d = prim::GetAttr[name="norm"](%self.1)
  %conv : __torch__.torch.nn.modules.conv.___torch_mangle_3.Conv2d = prim::GetAttr[name="conv"](%self.1)
  %55 : Tensor = prim::CallMethod[name="forward"](%conv, %x)
  %36 : Long(requires_grad=0, device=cpu) = prim::Constant[value={0}]() # convert_trace_in_script_with_if.py:18:0
  %37 : int = prim::Constant[value=1]() # convert_trace_in_script_with_if.py:18:0
  %input : Float(1, 16, 100, 100, strides=[160000, 10000, 100, 1], requires_grad=1, device=cpu) = aten::add(%55, %36, %37) # convert_trace_in_script_with_if.py:18:0
  %56 : Tensor = prim::CallMethod[name="forward"](%norm, %input)
  return (%56)

graph(%self : __torch__.TraceInScriptWithIf,
      %x.1 : Tensor):
  %3 : NoneType = prim::Constant()
  %5 : int = prim::Constant[value=0]() # convert_trace_in_script_with_if.py:32:21
  %4 : Tensor = aten::sum(%x.1, %3) # convert_trace_in_script_with_if.py:32:11
  %6 : Tensor = aten::gt(%4, %5) # convert_trace_in_script_with_if.py:32:11
  %8 : bool = aten::Bool(%6) # convert_trace_in_script_with_if.py:32:11
  %31 : Tensor = prim::If(%8) # convert_trace_in_script_with_if.py:32:8
    block0():
      %traced_conv_one_block : __torch__.ConvBlock = prim::GetAttr[name="traced_conv_one_block"](%self)
      %11 : Tensor = prim::CallMethod[name="forward"](%traced_conv_one_block, %x.1) # convert_trace_in_script_with_if.py:33:19
      -> (%11)
    block1():
      %traced_conv_two.1 : __torch__.___torch_mangle_5.ConvBlock = prim::GetAttr[name="traced_conv_two"](%self)
      %29 : Tensor = prim::CallMethod[name="forward"](%traced_conv_two.1, %x.1) # convert_trace_in_script_with_if.py:34:15
      -> (%29)
  return (%31)

graph torch-jit-export (
  %input[FLOAT, nx3xhxw]
) initializers (
  %traced_conv_one_block.conv.weight[FLOAT, 16x3x3x3]
  %traced_conv_one_block.norm.running_var[FLOAT, 16]
  %traced_conv_one_block.norm.running_mean[FLOAT, 16]
  %traced_conv_one_block.norm.bias[FLOAT, 16]
  %traced_conv_one_block.norm.weight[FLOAT, 16]
  %traced_conv_two.conv.weight[FLOAT, 16x3x3x3]
  %traced_conv_two.norm.running_var[FLOAT, 16]
  %traced_conv_two.norm.running_mean[FLOAT, 16]
  %traced_conv_two.norm.bias[FLOAT, 16]
  %traced_conv_two.norm.weight[FLOAT, 16]
) {
  %11 = ReduceSum[keepdims = 0](%input)
  %12 = Constant[value = <Scalar Tensor []>]()
  %13 = Greater(%11, %12)
  %14 = Cast[to = 9](%13)
  %output = If[else_branch = <graph torch-jit-export2>, then_branch = <graph torch-jit-export1>](%14)
  return %output
}

graph torch-jit-export1 {
  %16 = Conv[dilations = [1, 1], group = 1, kernel_shape = [3, 3], pads = [1, 1, 1, 1], strides = [1, 1]](%input, %traced_conv_one_block.conv.weight)
  %17 = Constant[value = <Scalar Tensor []>]()
  %18 = Add(%16, %17)
  %19 = BatchNormalization[epsilon = 9.99999974737875e-06, momentum = 0.899999976158142](%18, %traced_conv_one_block.norm.weight, %traced_conv_one_block.norm.bias, %traced_conv_one_block.norm.running_mean, %traced_conv_one_block.norm.running_var)
  return %19
}

graph torch-jit-export2 {
  %20 = Conv[dilations = [1, 1], group = 1, kernel_shape = [3, 3], pads = [1, 1, 1, 1], strides = [1, 1]](%input, %traced_conv_two.conv.weight)
  %21 = Constant[value = <Scalar Tensor []>]()
  %22 = Add(%20, %21)
  %23 = BatchNormalization[epsilon = 9.99999974737875e-06, momentum = 0.899999976158142](%22, %traced_conv_two.norm.weight, %traced_conv_two.norm.bias, %traced_conv_two.norm.running_mean, %traced_conv_two.norm.running_var)
  return %23
}
2021-11-22 17:00:26.552281291 [W:onnxruntime:, graph.cc:3391 CleanUnusedInitializers] Removing initializer '21'. It is not used by any node and should be removed from the model.
2021-11-22 17:00:26.552302936 [W:onnxruntime:, graph.cc:3391 CleanUnusedInitializers] Removing initializer '17'. It is not used by any node and should be removed from the model.
Check ort inference
ort_zeros_out OK
```

## Expected behavior

The exported onnx graph should be valid.

## Environment

PyTorch version: 1.10.0
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.8.12 (default, Oct 12 2021, 13:49:34)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.11.0-40-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.0.221
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1650 Ti
Nvidia driver version: 470.82.00
cuDNN version: Probably one of the following:
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.2
[pip3] pytorch-lightning==1.4.9
[pip3] torch==1.10.0
[pip3] torchaudio==0.10.0
[pip3] torchmetrics==0.6.0
[pip3] torchvision==0.11.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.1.74              h6bb024c_0    nvidia
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.3.0           h06a4308_520  
[conda] mkl-service               2.4.0            py38h7f8727e_0  
[conda] mkl_fft                   1.3.1            py38hd3c417c_0  
[conda] mkl_random                1.2.2            py38h51133e4_0  
[conda] mypy-extensions           0.4.3                    pypi_0    pypi
[conda] numpy                     1.19.0                   pypi_0    pypi
[conda] numpy-base                1.21.2           py38h79a1101_0  
[conda] pytorch                   1.10.0          py3.8_cuda11.1_cudnn8.0.5_0    pytorch
[conda] pytorch-lightning         1.4.9                    pypi_0    pypi
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.10.0               py38_cu111    pytorch
[conda] torchmetrics              0.6.0                    pypi_0    pypi
[conda] torchvision               0.11.1               py38_cu111    pytorch

**Comment by BowenBao** [comment]
hi @waihokwok thanks for reporting. I believe this issue is already fixed via #67272. Verified successful run of the script with torch nightly.

**Comment by shubhambhokare1** [comment]
Hi @waihokwok,

We’ve gone ahead and closed this issue because it has been resolved elsewhere. This resolution can be found [here](https://github.com/pytorch/pytorch/pull/67272).
If you still believe this issue is relevant, please feel free to reopen the issue and we will triage it as necessary. Please specify in a comment any updated information you may have so that we can address it effectively. We encourage you to try the latest pytorch-preview (nightly) version to see if it has resolved the issue.

Thanks,
ONNX Converter team
